### ISSUE TITLE
Should I create separate issues for testing and implementation?

### ISSUE THREAD

**Issue by inf3rno** [inf3rno/patterns#44]
Currently I write tests if I decided something about the API, but I don't wait the implementation, I close the issue after that. Should I start a separate issue about the implementation or should I use the same and use some sort of flag to indicate that the tests are already written, but it is not implemented yet. We could add a status flag for it, e.g. tested, pending, etc.

**Comment by inf3rno** [comment]
I did not want another label for this, but maybe that is the clearest solution. I don't want separate issues for implementation and tests. Better to have lifecycle flags: tested, implemented, documented. I use to close issues after I have tests, because implementing it is a lot easier than thinking about the API and writing the tests. We could close the issue after the tests were added to the codebase and add the implemented flag when we implement it and add the documented flag when we document it. Not a standard approach, but I do it alone.

Hmm better to be explicit here and add a tested flag and keep it open. We can close it after implementation and we can add a documented flag to the closed issue. I'll create the labels for this and reopen the old issues I closed after writing the tests.

**Comment by inf3rno** [comment]
Hmm another option would be to have an issue to track implementation and documentation. Actually what I really want is opening the issues with a bookmark and having a short list of the tasks I have to do. I don't think the current 1:1 milestone - major version system is a good one. The "Task Queue" is fine. I think the other should be the "Next Release" and we could remove the issues from it after we released. Keeping the issues in their milestone by version is not necessarily useful. Actually I never used that to find an issue and I doubt anybody else does it frequently. This is just a small change. It would be really useful to use these milestones as folders instead and have a tested and implemented milestone. With that approach we would have the following milestones:
 - later
 - design
 - test
 - implement
 - document
Actually these can be status flags too. So either we use labels or milestones for them. Label does not feel right because they exclude each other. On the other hand we can have something in the later milestone that is already tested, but I want to implement it later. We could write the documentation before the tests and implementation too, so in 1% of the cases these flags don't exclude each other.

Ok. I keep the current milestone system and add these progress flags. Currently we have tested and documented flags, I'll change that to do:design, do:test, do:document, do:implement. Better to ad what is needed instead of what is done.

**Comment by inf3rno** [comment]
Ok, I added the new labels, now it looks ok. I just need to check the old issues and reopen what is not implemented. Actually I don't really need to reopen them, I just need to add the "do" labels to them, so I can find what I need to work on later. An issue can be closed after the design and test part is done. The implementation and documentation is just formality in most of the cases.

**Comment by inf3rno** [comment]
I think the best way to add the do labels to all the open issues and check the commits of the tests and find the closed, but not implemented issues that way.

**Comment by inf3rno** [comment]
I thought about it, and I think I should keep the issues open until they are implemented and documented.

**Comment by inf3rno** [comment]
I think I should use different colors for the do: implement, document than I use for the do: research, design, test. The design and test is the real work, so that should be highlighted e.g. with red or orange. With that I won't need the priority flag. I think I'll remove the invalid flag too. It does not make sense. I mean who will look for invalid issues?! Better to write a text about why I close it and just close it without any flag.

**Comment by inf3rno** [comment]
Ok. I'll add the new labels to the issues and reopen the not yet implemented and documented issues.

**Comment by inf3rno** [comment]
I was not sure if I should add all missing do flags or just the one that indicates the actual status. I decided to add only a single flag, so research/design/implement/document. The order is the same, so I'll document a feature only after implementation. That is in contrast with literal programming, but I can live with that. I separated the problem descriptions and the workflow from the documentation. I use the issues for the former and the wiki for the latter. I document only the results and not the goals and decision making. I'll remove the "decisions about the lib" part from the wiki. The whys can be found in the issues, and they aren't easy to document. Maybe I should add some texts about the whys to the documentation of each features. The same is true for the research. I don't think I should fully document researches separately. I should mention the results in the feature descriptions of the documentation.

**Comment by inf3rno** [comment]
I don't think I'll document meta decisions.

**Comment by inf3rno** [comment]
I think we should have only a single do flag for an issue. Whenever we have multiple ones, the issue might be too big or we are doing things in the wrong order. I think better to document this.
